File: streampark-flink/streampark-flink-shims/streampark-flink-shims-base/src/main/scala/org/apache/streampark/flink/core/FlinkTableTrait.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.core

import org.apache.streampark.common.util.Utils
import org.apache.streampark.flink.core.EnhancerImplicit._

import org.apache.flink.api.common.JobExecutionResult
import org.apache.flink.api.java.utils.ParameterTool
import org.apache.flink.table.api._
import org.apache.flink.table.catalog.Catalog
import org.apache.flink.table.expressions.Expression
import org.apache.flink.table.functions._
import org.apache.flink.table.module.Module
import org.apache.flink.table.types.AbstractDataType

import java.lang
import java.util.Optional

abstract class FlinkTableTrait(val parameter: ParameterTool, private val tableEnv: TableEnvironment)
  extends TableEnvironment {

  def start(): JobExecutionResult = {
    val appName = parameter.getAppName(required = true)
    execute(appName)
  }

  def execute(jobName: String): JobExecutionResult = {
    Utils.printLogo(s"FlinkTable $jobName Starting...")
    null
  }

  def sql(sql: String = null): Unit =
    FlinkSqlExecutor.executeSql(sql, parameter, this)

  override def fromValues(values: Expression*): Table =
    tableEnv.fromValues(values)

  override def fromValues(rowType: AbstractDataType[_], values: Expression*): Table =
    tableEnv.fromValues(rowType, values: _*)

  override def fromValues(values: lang.Iterable[_]): Table =
    tableEnv.fromValues(values)

  override def fromValues(rowType: AbstractDataType[_], values: lang.Iterable[_]): Table =
    tableEnv.fromValues(rowType, values)

  override def registerCatalog(catalogName: String, catalog: Catalog): Unit =
    tableEnv.registerCatalog(catalogName, catalog)

  override def getCatalog(catalogName: String): Optional[Catalog] =
    tableEnv.getCatalog(catalogName)

  override def loadModule(moduleName: String, module: Module): Unit =
    tableEnv.loadModule(moduleName, module)

  override def unloadModule(moduleName: String): Unit =
    tableEnv.unloadModule(moduleName)

  override def createTemporarySystemFunction(
      name: String,
      functionClass: Class[_ <: UserDefinedFunction]): Unit =
    tableEnv.createTemporarySystemFunction(name, functionClass)

  override def createTemporarySystemFunction(
      name: String,
      functionInstance: UserDefinedFunction): Unit =
    tableEnv.createTemporarySystemFunction(name, functionInstance)

  override def dropTemporarySystemFunction(name: String): Boolean =
    tableEnv.dropTemporarySystemFunction(name)

  override def createFunction(path: String, functionClass: Class[_ <: UserDefinedFunction]): Unit =
    tableEnv.createFunction(path, functionClass)

  override def createFunction(
      path: String,
      functionClass: Class[_ <: UserDefinedFunction],
      ignoreIfExists: Boolean): Unit =
    tableEnv.createFunction(path, functionClass)

  override def dropFunction(path: String): Boolean = tableEnv.dropFunction(path)

  override def createTemporaryFunction(
      path: String,
      functionClass: Class[_ <: UserDefinedFunction]): Unit =
    tableEnv.createTemporaryFunction(path, functionClass)

  override def createTemporaryFunction(path: String, functionInstance: UserDefinedFunction): Unit =
    tableEnv.createTemporaryFunction(path, functionInstance)

  override def dropTemporaryFunction(path: String): Boolean =
    tableEnv.dropTemporaryFunction(path)

  override def createTemporaryView(path: String, view: Table): Unit =
    tableEnv.createTemporaryView(path, view)

  override def from(path: String): Table = tableEnv.from(path)

  override def listCatalogs(): Array[String] = tableEnv.listCatalogs()

  override def listModules(): Array[String] = tableEnv.listModules()

  override def listDatabases(): Array[String] = tableEnv.listDatabases()

  override def listTables(): Array[String] = tableEnv.listTables()

  override def listViews(): Array[String] = tableEnv.listViews()

  override def listTemporaryTables(): Array[String] =
    tableEnv.listTemporaryTables

  override def listTemporaryViews(): Array[String] =
    tableEnv.listTemporaryViews()

  override def listUserDefinedFunctions(): Array[String] =
    tableEnv.listUserDefinedFunctions()

  override def listFunctions(): Array[String] = tableEnv.listFunctions()

  override def dropTemporaryTable(path: String): Boolean =
    tableEnv.dropTemporaryTable(path)

  override def dropTemporaryView(path: String): Boolean =
    tableEnv.dropTemporaryView(path)

  override def explainSql(statement: String, extraDetails: ExplainDetail*): String =
    tableEnv.explainSql(statement, extraDetails: _*)

  override def sqlQuery(query: String): Table = tableEnv.sqlQuery(query)

  override def executeSql(statement: String): TableResult =
    tableEnv.executeSql(statement)

  override def getCurrentCatalog: String = tableEnv.getCurrentCatalog

  override def useCatalog(catalogName: String): Unit =
    tableEnv.useCatalog(catalogName)

  override def getCurrentDatabase: String = tableEnv.getCurrentDatabase

  override def useDatabase(databaseName: String): Unit =
    tableEnv.useDatabase(databaseName)

  override def getConfig: TableConfig = tableEnv.getConfig

  override def createStatementSet(): StatementSet =
    tableEnv.createStatementSet()

  @deprecated override def registerFunction(name: String, function: ScalarFunction): Unit =
    tableEnv.registerFunction(name, function)

  @deprecated override def registerTable(name: String, table: Table): Unit =
    tableEnv.registerTable(name, table)

  @deprecated override def scan(tablePath: String*): Table =
    tableEnv.scan(tablePath: _*)

  @deprecated override def getCompletionHints(statement: String, position: Int): Array[String] =
    tableEnv.getCompletionHints(statement, position)
}

File: streampark-flink/streampark-flink-shims/streampark-flink-shims-base/src/main/java/org/apache/streampark/flink/core/FlinkTableTrait.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.core;

import org.apache.streampark.common.util.Utils;

import org.apache.flink.api.common.JobExecutionResult;
import org.apache.flink.api.java.utils.ParameterTool;
import org.apache.flink.table.api.ExplainDetail;
import org.apache.flink.table.api.StatementSet;
import org.apache.flink.table.api.Table;
import org.apache.flink.table.api.TableConfig;
import org.apache.flink.table.api.TableEnvironment;
import org.apache.flink.table.api.TableResult;
import org.apache.flink.table.catalog.Catalog;
import org.apache.flink.table.expressions.Expression;
import org.apache.flink.table.functions.ScalarFunction;
import org.apache.flink.table.functions.UserDefinedFunction;
import org.apache.flink.table.module.Module;
import org.apache.flink.table.types.AbstractDataType;

import java.util.Optional;

public abstract class FlinkTableTrait implements TableEnvironment {

    public final ParameterTool parameter;
    private final TableEnvironment tableEnv;

    protected FlinkTableTrait(ParameterTool parameter, TableEnvironment tableEnv) {
        this.parameter = parameter;
        this.tableEnv = tableEnv;
    }

    protected TableEnvironment delegate() {
        return tableEnv;
    }

    public JobExecutionResult start() {
        String appName = FlinkEnvironmentUtils.getAppName(parameter, null, true);
        return execute(appName);
    }

    public JobExecutionResult execute(String jobName) {
        Utils.printLogo("FlinkTable " + jobName + " Starting...");
        return null;
    }

    public void sql() {
        sql(null);
    }

    public void sql(String sql) {
        FlinkSqlExecutor.executeSql(sql, parameter, this);
    }

    @Override
    public Table fromValues(Expression... values) {
        return delegate().fromValues(values);
    }

    @Override
    public Table fromValues(AbstractDataType<?> rowType, Expression... values) {
        return delegate().fromValues(rowType, values);
    }

    @Override
    public Table fromValues(Iterable<?> values) {
        return delegate().fromValues(values);
    }

    @Override
    public Table fromValues(AbstractDataType<?> rowType, Iterable<?> values) {
        return delegate().fromValues(rowType, values);
    }

    @Override
    public void registerCatalog(String catalogName, Catalog catalog) {
        delegate().registerCatalog(catalogName, catalog);
    }

    @Override
    public Optional<Catalog> getCatalog(String catalogName) {
        return delegate().getCatalog(catalogName);
    }

    @Override
    public void loadModule(String moduleName, Module module) {
        delegate().loadModule(moduleName, module);
    }

    @Override
    public void unloadModule(String moduleName) {
        delegate().unloadModule(moduleName);
    }

    @Override
    public void createTemporarySystemFunction(
                                              String name, Class<? extends UserDefinedFunction> functionClass) {
        delegate().createTemporarySystemFunction(name, functionClass);
    }

    @Override
    public void createTemporarySystemFunction(String name, UserDefinedFunction functionInstance) {
        delegate().createTemporarySystemFunction(name, functionInstance);
    }

    @Override
    public boolean dropTemporarySystemFunction(String name) {
        return delegate().dropTemporarySystemFunction(name);
    }

    @Override
    public void createFunction(String path, Class<? extends UserDefinedFunction> functionClass) {
        delegate().createFunction(path, functionClass);
    }

    @Override
    public void createFunction(
                               String path,
                               Class<? extends UserDefinedFunction> functionClass,
                               boolean ignoreIfExists) {
        delegate().createFunction(path, functionClass, ignoreIfExists);
    }

    @Override
    public boolean dropFunction(String path) {
        return delegate().dropFunction(path);
    }

    @Override
    public void createTemporaryFunction(String path, Class<? extends UserDefinedFunction> functionClass) {
        delegate().createTemporaryFunction(path, functionClass);
    }

    @Override
    public void createTemporaryFunction(String path, UserDefinedFunction functionInstance) {
        delegate().createTemporaryFunction(path, functionInstance);
    }

    @Override
    public boolean dropTemporaryFunction(String path) {
        return delegate().dropTemporaryFunction(path);
    }

    @Override
    public void createTemporaryView(String path, Table view) {
        delegate().createTemporaryView(path, view);
    }

    @Override
    public Table from(String path) {
        return delegate().from(path);
    }

    @Override
    public String[] listCatalogs() {
        return delegate().listCatalogs();
    }

    @Override
    public String[] listModules() {
        return delegate().listModules();
    }

    @Override
    public String[] listDatabases() {
        return delegate().listDatabases();
    }

    @Override
    public String[] listTables() {
        return delegate().listTables();
    }

    @Override
    public String[] listViews() {
        return delegate().listViews();
    }

    @Override
    public String[] listTemporaryTables() {
        return delegate().listTemporaryTables();
    }

    @Override
    public String[] listTemporaryViews() {
        return delegate().listTemporaryViews();
    }

    @Override
    public String[] listUserDefinedFunctions() {
        return delegate().listUserDefinedFunctions();
    }

    @Override
    public String[] listFunctions() {
        return delegate().listFunctions();
    }

    @Override
    public boolean dropTemporaryTable(String path) {
        return delegate().dropTemporaryTable(path);
    }

    @Override
    public boolean dropTemporaryView(String path) {
        return delegate().dropTemporaryView(path);
    }

    @Override
    public String explainSql(String statement, ExplainDetail... extraDetails) {
        return delegate().explainSql(statement, extraDetails);
    }

    @Override
    public Table sqlQuery(String query) {
        return delegate().sqlQuery(query);
    }

    @Override
    public TableResult executeSql(String statement) {
        return delegate().executeSql(statement);
    }

    @Override
    public String getCurrentCatalog() {
        return delegate().getCurrentCatalog();
    }

    @Override
    public void useCatalog(String catalogName) {
        delegate().useCatalog(catalogName);
    }

    @Override
    public String getCurrentDatabase() {
        return delegate().getCurrentDatabase();
    }

    @Override
    public void useDatabase(String databaseName) {
        delegate().useDatabase(databaseName);
    }

    @Override
    public TableConfig getConfig() {
        return delegate().getConfig();
    }

    @Override
    public StatementSet createStatementSet() {
        return delegate().createStatementSet();
    }

    @Deprecated
    @Override
    public void registerFunction(String name, ScalarFunction function) {
        delegate().registerFunction(name, function);
    }

    @Deprecated
    @Override
    public void registerTable(String name, Table table) {
        delegate().registerTable(name, table);
    }

    @Deprecated
    @Override
    public Table scan(String... tablePath) {
        return delegate().scan(tablePath);
    }

    @Deprecated
    @Override
    public String[] getCompletionHints(String statement, int position) {
        return delegate().getCompletionHints(statement, position);
    }
}
